python/hail/expr/expressions/base_expression.py

```diff
@@ -511,20 +446,20 @@ def _slice(self, ret_type, start=None, stop=None, step=None):
             raise NotImplementedError('Variable slice step size is not currently supported')
 
         non_null = [x for x in [start, stop] if x is not None]
-        indices, aggregations, joins, refs = unify_all(self, *non_null)
+        indices, aggregations, joins = unify_all(self, *non_null)
         return expressions.construct_expr(Index(self._ast, Slice(start_ast, stop_ast)),
-                                          ret_type, indices, aggregations, joins, refs)
+                                          ret_type, indices, aggregations, joins)
 
     def _bin_lambda_method(self, name, f, input_type, ret_type_f, *args):
         args = (to_expr(arg) for arg in args)
         new_id = Env._get_uid()
         lambda_result = to_expr(
-            f(expressions.construct_expr(Reference(new_id), input_type, self._indices, self._aggregations, self._joins,
-                                         self._refs)))
+            f(expressions.construct_expr(VariableReference(new_id), input_type, self._indices, self._aggregations,
```

Comment:
Is there a reason why the constructor to VariableReference doesn't create the new `uid`?

Reply:
not sure, it might be OK here. Let's change that in a separate PR, this one is big enough.